File: wizards/com/sun/star/wizards/text/TextFieldHandler.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */
package com.sun.star.wizards.text;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.GregorianCalendar;

import com.sun.star.beans.XPropertySet;
import com.sun.star.container.XEnumeration;
import com.sun.star.lang.XComponent;
import com.sun.star.lang.XMultiServiceFactory;
import com.sun.star.lang.XServiceInfo;
import com.sun.star.text.XDependentTextField;
import com.sun.star.text.XTextContent;
import com.sun.star.text.XTextCursor;
import com.sun.star.text.XTextDocument;
import com.sun.star.text.XTextFieldsSupplier;
import com.sun.star.text.XTextRange;
import com.sun.star.uno.AnyConverter;
import com.sun.star.uno.Exception;
import com.sun.star.uno.UnoRuntime;
import com.sun.star.uno.XInterface;
import com.sun.star.util.DateTime;
import com.sun.star.util.XRefreshable;
import com.sun.star.util.XUpdatable;
import com.sun.star.wizards.common.Helper;
import com.sun.star.wizards.common.PropertyNames;

public class TextFieldHandler
{

    public XTextFieldsSupplier xTextFieldsSupplier;
    private XMultiServiceFactory xMSFDoc;

    /**
     * Creates a new instance of TextFieldHandler
     * @param xMSF
     * @param xTextDocument
     */
    public TextFieldHandler(XMultiServiceFactory xMSF, XTextDocument xTextDocument)
    {
        this.xMSFDoc = xMSF;
        xTextFieldsSupplier = UnoRuntime.queryInterface(XTextFieldsSupplier.class, xTextDocument);
    }

    public void refreshTextFields()
    {
        XRefreshable xUp = UnoRuntime.queryInterface(XRefreshable.class, xTextFieldsSupplier.getTextFields());
        xUp.refresh();
    }

    public String getUserFieldContent(XTextCursor xTextCursor)
    {
        try
        {
            XTextRange xTextRange = xTextCursor.getEnd();
            Object oTextField = Helper.getUnoPropertyValue(xTextRange, "TextField");
            if (com.sun.star.uno.AnyConverter.isVoid(oTextField))
            {
                return PropertyNames.EMPTY_STRING;
            }
            else
            {
                XDependentTextField xDependent = UnoRuntime.queryInterface(XDependentTextField.class, oTextField);
                XPropertySet xMaster = xDependent.getTextFieldMaster();
                return (String) xMaster.getPropertyValue("Content");
            }
        }
        catch (com.sun.star.uno.Exception exception)
        {
            exception.printStackTrace(System.err);
        }
        return PropertyNames.EMPTY_STRING;
    }

    public void insertUserField(XTextCursor xTextCursor, String FieldName, String FieldTitle)
    {
        try
        {
            XInterface xField = (XInterface) xMSFDoc.createInstance("com.sun.star.text.TextField.User");
            XDependentTextField xDepField = UnoRuntime.queryInterface(XDependentTextField.class, xField);
            XTextContent xFieldContent = UnoRuntime.queryInterface(XTextContent.class, xField);
            if (xTextFieldsSupplier.getTextFieldMasters().hasByName("com.sun.star.text.FieldMaster.User." + FieldName))
            {
                Object oMaster = xTextFieldsSupplier.getTextFieldMasters().getByName("com.sun.star.text.FieldMaster.User." + FieldName);
                XComponent xComponent = UnoRuntime.queryInterface(XComponent.class, oMaster);
                xComponent.dispose();
            }
            XPropertySet xPSet = createUserField(FieldName, FieldTitle);
            xDepField.attachTextFieldMaster(xPSet);
            xTextCursor.getText().insertTextContent(xTextCursor, xFieldContent, false);

//            try
//            {
//                XPropertySet xTestProp = xDepField.getTextFieldMaster();
//                String UserFieldName = (String) xTestProp.getPropertyValue(PropertyNames.PROPERTY_NAME);
//                // UserFieldName == FieldName?
//                int dummy = 0;
//            }
//            catch (com.sun.star.uno.Exception e)
//            {
//                int dummy2 = 0;
//            }

        }
        catch (com.sun.star.uno.Exception exception)
        {
            exception.printStackTrace(System.err);
        }
    }

    public XPropertySet createUserField(String FieldName, String FieldTitle) throws com.sun.star.uno.Exception
    {
        Object oMaster = xMSFDoc.createInstance("com.sun.star.text.FieldMaster.User");
        XPropertySet xPSet = UnoRuntime.queryInterface(XPropertySet.class, oMaster);
        xPSet.setPropertyValue(PropertyNames.PROPERTY_NAME, FieldName);
        xPSet.setPropertyValue("Content", FieldTitle);

        // DEBUG
        // String sFieldName = (String)xPSet.getPropertyValue(PropertyNames.PROPERTY_NAME);

        return xPSet;
    }

    private XDependentTextField[] getTextFieldsByProperty(String _PropertyName, Object _aPropertyValue, String _TypeName) throws Exception
    {
        try
        {
            XDependentTextField[] xDependentFields;
            ArrayList<XDependentTextField> xDependentVector = new ArrayList<XDependentTextField>();
            if (xTextFieldsSupplier.getTextFields().hasElements())
            {
                XEnumeration xEnum = xTextFieldsSupplier.getTextFields().createEnumeration();
                while (xEnum.hasMoreElements())
                {
                    Object oTextField = xEnum.nextElement();
                    XDependentTextField xDependent = UnoRuntime.queryInterface(XDependentTextField.class, oTextField);
                    XPropertySet xPropertySet = xDependent.getTextFieldMaster();
                    if (xPropertySet.getPropertySetInfo().hasPropertyByName(_PropertyName))
                    {
                        Object oValue = xPropertySet.getPropertyValue(_PropertyName);
                        // TODO replace the following comparison via com.sun.star.uno.Any.Type
                        if (AnyConverter.isString(oValue))
                        {
                            if (_TypeName.equals("String"))
                            {
                                String sValue = AnyConverter.toString(oValue);
                                if (sValue.equals(_aPropertyValue))
                                {
                                    xDependentVector.add(xDependent);
                                }
                            }
                        }
                        else if (AnyConverter.isShort(oValue))
                        {
                            if (_TypeName.equals("Short"))
                            {
                                short iShortParam = ((Short) _aPropertyValue).shortValue();
                                short ishortValue = AnyConverter.toShort(oValue);
                                if (ishortValue == iShortParam)
                                {
                                    xDependentVector.add(xDependent);
                                }
                            }
                        }
                    }
                }
            }
            if (xDependentVector.size() > 0)
            {
                xDependentFields = new XDependentTextField[xDependentVector.size()];
                xDependentVector.toArray(xDependentFields);
                return xDependentFields;
            }
        }
        catch (Exception e)
        {
            // TODO Auto-generated catch block
            e.printStackTrace(System.err);
        }
        return null;
    }

    public void changeUserFieldContent(String _FieldName, String _FieldContent)
    {
        try
        {
            XDependentTextField[] xDependentTextFields = getTextFieldsByProperty(PropertyNames.PROPERTY_NAME, _FieldName, "String");
            if (xDependentTextFields != null)
            {
                for (int i = 0; i < xDependentTextFields.length; i++)
                {
                    xDependentTextFields[i].getTextFieldMaster().setPropertyValue("Content", _FieldContent);
                }
                refreshTextFields();
            }
        }
        catch (Exception e)
        {
            e.printStackTrace(System.err);
        }
    }

    public void updateDocInfoFields()
    {
        try
        {
            XEnumeration xEnum = xTextFieldsSupplier.getTextFields().createEnumeration();
            while (xEnum.hasMoreElements())
            {
                Object oTextField = xEnum.nextElement();
                XServiceInfo xSI = UnoRuntime.queryInterface(XServiceInfo.class, oTextField);

                if (xSI.supportsService("com.sun.star.text.TextField.ExtendedUser"))
                {
                    XUpdatable xUp = UnoRuntime.queryInterface(XUpdatable.class, oTextField);
                    xUp.update();
                }
                if (xSI.supportsService("com.sun.star.text.TextField.User"))
                {
                    XUpdatable xUp = UnoRuntime.queryInterface(XUpdatable.class, oTextField);
                    xUp.update();
                }
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    public void updateDateFields()
    {
        try
        {
            XEnumeration xEnum = xTextFieldsSupplier.getTextFields().createEnumeration();
            Calendar cal = new GregorianCalendar();
            DateTime dt = new DateTime();
            dt.Day = (short) cal.get(Calendar.DAY_OF_MONTH);
            dt.Year = (short) cal.get(Calendar.YEAR);
            dt.Month = (short) cal.get(Calendar.MONTH);
            dt.Month++;

            while (xEnum.hasMoreElements())
            {
                Object oTextField = xEnum.nextElement();
                XServiceInfo xSI = UnoRuntime.queryInterface(XServiceInfo.class, oTextField);

                if (xSI.supportsService("com.sun.star.text.TextField.DateTime"))
                {
                    XPropertySet xPSet = UnoRuntime.queryInterface(XPropertySet.class, oTextField);
                    xPSet.setPropertyValue("IsFixed", Boolean.FALSE);
                    xPSet.setPropertyValue("DateTimeValue", dt);
                }
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    public void fixDateFields(boolean _bSetFixed)
    {
        try
        {
            XEnumeration xEnum = xTextFieldsSupplier.getTextFields().createEnumeration();
            while (xEnum.hasMoreElements())
            {
                Object oTextField = xEnum.nextElement();
                XServiceInfo xSI = UnoRuntime.queryInterface(XServiceInfo.class, oTextField);
                if (xSI.supportsService("com.sun.star.text.TextField.DateTime"))
                {
                    XPropertySet xPSet = UnoRuntime.queryInterface(XPropertySet.class, oTextField);
                    xPSet.setPropertyValue("IsFixed", Boolean.valueOf(_bSetFixed));
                }
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    public void removeUserFieldByContent(String _FieldContent)
    {
        try
        {
            XDependentTextField[] xDependentTextFields = getTextFieldsByProperty("Content", _FieldContent, "String");
            if (xDependentTextFields != null)
            {
                for (int i = 0; i < xDependentTextFields.length; i++)
                {
                    xDependentTextFields[i].dispose();
                }
            }
        }
        catch (Exception e)
        {
            e.printStackTrace(System.err);
        }
    }

    public void changeExtendedUserFieldContent(short UserDataPart, String _FieldContent)
    {
        try
        {
            XDependentTextField[] xDependentTextFields = getTextFieldsByProperty("UserDataType", new Short(UserDataPart), "Short");
            if (xDependentTextFields != null)
            {
                for (int i = 0; i < xDependentTextFields.length; i++)
                {
                    xDependentTextFields[i].getTextFieldMaster().setPropertyValue("Content", _FieldContent);
                }
            }
            refreshTextFields();
        }
        catch (Exception e)
        {
            e.printStackTrace(System.err);
        }
    }
}

File: wizards/com/sun/star/wizards/text/TextFieldHandler.py
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# This file incorporates work covered by the following license notice:
#
#   Licensed to the Apache Software Foundation (ASF) under one or more
#   contributor license agreements. See the NOTICE file distributed
#   with this work for additional information regarding copyright
#   ownership. The ASF licenses this file to you under the Apache
#   License, Version 2.0 (the "License"); you may not use this file
#   except in compliance with the License. You may obtain a copy of
#   the License at http://www.apache.org/licenses/LICENSE-2.0 .
#
import traceback
import time
import unicodedata
from wizards.common.PropertyNames import PropertyNames

from com.sun.star.util import DateTime

class TextFieldHandler(object):
    '''
    Creates a new instance of TextFieldHandler
    @param xMSF
    @param xTextDocument
    '''

    xTextFieldsSupplierAux = None
    arrayTextFields = None
    dictTextFields = None

    def __init__(self, xMSF, xTextDocument):
        self.xMSFDoc = xMSF
        self.xTextFieldsSupplier = xTextDocument
        if TextFieldHandler.xTextFieldsSupplierAux is not \
                self.xTextFieldsSupplier:
            self.__getTextFields()
            TextFieldHandler.xTextFieldsSupplierAux = self.xTextFieldsSupplier

    def refreshTextFields(self):
        xUp = self.xTextFieldsSupplier.TextFields
        xUp.refresh()

    def getUserFieldContent(self, xTextCursor):
        try:
            xTextRange = xTextCursor.getEnd()
            oTextField = Helper.getUnoPropertyValue(xTextRange, "TextField")
            if com.sun.star.uno.AnyConverter.isVoid(oTextField):
                return ""
            else:
                xMaster = oTextField.getTextFieldMaster()
                UserFieldContent = xMaster.getPropertyValue("Content")
                return UserFieldContent

        except Exception, exception:
            traceback.print_exc()

        return ""

    def insertUserField(self, xTextCursor, FieldName, FieldTitle):
        try:
            xField = self.xMSFDoc.createInstance(
                "com.sun.star.text.TextField.User")

            if self.xTextFieldsSupplier.TextFieldMasters.hasByName(
                "com.sun.star.text.FieldMaster.User." + FieldName):
                oMaster = self.xTextFieldsSupplier.TextFieldMasters.getByName( \
                    "com.sun.star.text.FieldMaster.User." + FieldName)
                oMaster.dispose()

            xPSet = createUserField(FieldName, FieldTitle)
            xField.attachTextFieldMaster(xPSet)
            xTextCursor.getText().insertTextContent(
                xTextCursor, xField, False)
        except com.sun.star.uno.Exception, exception:
            traceback.print_exc()

    def createUserField(self, FieldName, FieldTitle):
        xPSet = self.xMSFDoc.createInstance(
            "com.sun.star.text.FieldMaster.User")
        xPSet.setPropertyValue(PropertyNames.PROPERTY_NAME, FieldName)
        xPSet.setPropertyValue("Content", FieldTitle)
        return xPSet

    def __getTextFields(self):
        try:
            if self.xTextFieldsSupplier.TextFields.hasElements():
                TextFieldHandler.dictTextFields = {}
                TextFieldHandler.arrayTextFields = []
                xEnum = \
                    self.xTextFieldsSupplier.TextFields.createEnumeration()
                while xEnum.hasMoreElements():
                    oTextField = xEnum.nextElement()
                    TextFieldHandler.arrayTextFields.append(oTextField)
                    xPropertySet = oTextField.TextFieldMaster
                    if xPropertySet.Name:
                        TextFieldHandler.dictTextFields[xPropertySet.Name] = \
                            oTextField
        except Exception, e:
            #TODO Auto-generated catch block
            traceback.print_exc()

    def __getTextFieldsByProperty(
            self, _PropertyName, _aPropertyValue):
        try:
            xProperty = TextFieldHandler.dictTextFields[_aPropertyValue]
            xPropertySet = xProperty.TextFieldMaster
            if xPropertySet.PropertySetInfo.hasPropertyByName(
                    _PropertyName):
                oValue = xPropertySet.getPropertyValue(_PropertyName)
                sValue = unicodedata.normalize(
                    'NFKD', oValue).encode('ascii','ignore')
                if sValue == _aPropertyValue:
                    return xProperty
            return None
        except KeyError, e:
            return None

    def changeUserFieldContent(self, _FieldName, _FieldContent):
        DependentTextFields = self.__getTextFieldsByProperty(
                PropertyNames.PROPERTY_NAME, _FieldName)
        if DependentTextFields is not None:
            DependentTextFields.TextFieldMaster.setPropertyValue("Content", _FieldContent)
            self.refreshTextFields()

    def updateDocInfoFields(self):
        try:
            for i in TextFieldHandler.arrayTextFields:
                if i.supportsService(
                    "com.sun.star.text.TextField.ExtendedUser"):
                    i.update()

                if i.supportsService(
                    "com.sun.star.text.TextField.User"):
                    i.update()

        except Exception, e:
            traceback.print_exc()

    def updateDateFields(self):
        try:
            now = time.localtime(time.time())
            dt = DateTime()
            dt.Day = time.strftime("%d", now)
            dt.Year = time.strftime("%Y", now)
            dt.Month = time.strftime("%m", now)
            dt.Month += 1
            for i in TextFieldHandler.arrayTextFields:
                if i.supportsService(
                    "com.sun.star.text.TextField.DateTime"):
                    i.setPropertyValue("IsFixed", False)
                    i.setPropertyValue("DateTimeValue", dt)

        except Exception, e:
            traceback.print_exc()

    def fixDateFields(self, _bSetFixed):
        try:
            for i in TextFieldHandler.arrayTextFields:
                if i.supportsService(
                    "com.sun.star.text.TextField.DateTime"):
                    i.setPropertyValue("IsFixed", _bSetFixed)

        except Exception, e:
            traceback.print_exc()

    def removeUserFieldByContent(self, _FieldContent):
        try:
            xDependentTextFields = self.__getTextFieldsByProperty(
                "Content", _FieldContent)
            if xDependentTextFields != None:
                i = 0
                while i < xDependentTextFields.length:
                    xDependentTextFields[i].dispose()
                    i += 1

        except Exception, e:
            traceback.print_exc()
